### DISCUSSION TITLE
tss signature can't verify success. but the key is right and I have verify in python

### DISCUSSION THREAD

**Discussion by xin19961512** [trustwallet/wallet-core#3602] [Q&A]
**I use TWPublicKeyRecover() to recover publickey;**
```
    auto signature = DATA("061b21a8ec032f1cfe7860bd2d31a028b17a1c83de285ad762dffc3f0e029adf69929b4c1639296cc7c077692a396cbd6656b8eb764b7defc339aa2738b6662001");
    auto hashData = DATA("0400a680400ef750d6813f96f0e80d919aa43416938e7a5dccfb82e9d42758ac");
WRAP(TWPublicKey, TWPublicKeyRecover(signature.get(), hashData.get());
```
but the address is wrong: 
from address:0x9f64152B97b3482f9dC20587aacd66dAE42B5985
recover address:0xd7e4b51f145ed9909AE2a2DF76F68a68a53ED743

**however, I can recovery it with python code :**
```
r_value = 0x061b21a8ec032f1cfe7860bd2d31a028b17a1c83de285ad762dffc3f0e029adf
s_value = 0x69929b4c1639296cc7c077692a396cbd6656b8eb764b7defc339aa2738b66620
v_value = 1

message_to_sign = [ord(char) for char in "0400a680400ef750d6813f96f0e80d919aa43416938e7a5dccfb82e9d42758ac"] 

signature = keys.Signature(vrs=(v, r, s))
recovered_public_key = signature.recover_public_key_from_msg_hash(message_to_sign)
recovered_address =recovered_public_key.to_address()

print(recovered_address)
```
it get the right address : 0x9f64152B97b3482f9dC20587aacd66dAE42B5985

by the way, this signature is created by tss. and I use wallet-core generate  signature is success . any suggestions?

**Comment by satoshiotomakan** [comment]
Hi @xin19961512, could you send a link to the "tss" you mentioned above?

**Reply by xin19961512** [reply]
https://github.com/ZenGo-X/multi-party-ecdsa

maybe my signature wrong? but, what kind of serialized signature does it need ?
//    signature { r:0xb89b778b5b6c6fe5a2f547d6d06b8712f4c4d9bfc7d525ac3a706be6cffae6e5, s:0x76179b9f4cd3a8e9cb3dd7f2268b42c758dbee40fd4728f5d1a83e0d882c2b53, v:1 }
auto signature = DATA("b89b778b5b6c6fe5a2f547d6d06b8712f4c4d9bfc7d525ac3a706be6cffae6e576179b9f4cd3a8e9cb3dd7f2268b42c758dbee40fd4728f5d1a83e0d882c2b5301");

**Reply by xin19961512** [reply]
hi  @satoshiotomakan , I use this signature， and print the RSV with SigningOutput. the recoverid = 35+chainid*2+1. so, I think my signature what I encode is right. but why I the recovered address is wrong?  thanks!
```c++
    //    signature { r:0xb89b778b5b6c6fe5a2f547d6d06b8712f4c4d9bfc7d525ac3a706be6cffae6e5, s:0x76179b9f4cd3a8e9cb3dd7f2268b42c758dbee40fd4728f5d1a83e0d882c2b53, v:1 }
    auto signature = DATA("b89b778b5b6c6fe5a2f547d6d06b8712f4c4d9bfc7d525ac3a706be6cffae6e576179b9f4cd3a8e9cb3dd7f2268b42c758dbee40fd4728f5d1a83e0d882c2b5301");

Ethereum::Proto::SigningOutput output;
    cout << "r:" << hex(output.r()) << endl
         << "s:" << hex(output.s()) <<endl
         << "v:" << hex(output.v())<< endl;
```
result is :
//r:b89b778b5b6c6fe5a2f547d6d06b8712f4c4d9bfc7d525ac3a706be6cffae6e5
//s:76179b9f4cd3a8e9cb3dd7f2268b42c758dbee40fd4728f5d1a83e0d882c2b53
//v:0a74

recover addresss with code 
```c++
    const auto recPubKey = WRAP(TWPublicKey, TWPublicKeyRecover(signature.get(), (TWData*)&preImageHashData));

    auto recaddress = WRAP(TWAnyAddress, TWAnyAddressCreateWithPublicKey(recPubKey.get(), coin));
    auto recaddressStr = WRAPS(TWAnyAddressDescription(recaddress.get()));
    auto strRecoveryAddr = TWStringUTF8Bytes(recaddressStr.get());
    cout << "recover address:"<< strRecoveryAddr << endl;

```
result is :0xB2f501220eb3c661042e940c2227748179ECB357
and the right address is :0x9f64152B97b3482f9dC20587aacd66dAE42B5985

**Comment by xin19961512** [comment]
i have solved it, thanks
